### PR TITLE
MAINT-1790 Allow pagination beyond 10K reference set members

### PR DESCRIPTION
**Pull request by dmcgihtsdo**

### CODE REVIEW

File: src/main/java/org/snomed/snowstorm/core/data/services/ReferenceSetMemberService.java [hunk 1]
```diff
@@ -60,6 +62,14 @@
...
 @Service
 public class ReferenceSetMemberService extends ComponentService {
+	private static final Function<ReferenceSetMember, Object[]> REFERENCE_SET_MEMBER_ID_SEARCH_AFTER_EXTRACTOR = referenceSetMember -> {
+		if (referenceSetMember == null) {
+			return null;
+		}
+
+		String id = referenceSetMember.getId();
+		return id == null ? null : SearchAfterHelper.convertToTokenAndBack(new Object[]{id});
+	};
```
Comment (dmcgihtsdo) [inline src/main/java/org/snomed/snowstorm/core/data/services/ReferenceSetMemberService.java:72]: I referenced other areas in the code that used SearchAfterPage. From what I can tell, the function here is used to determine which record to scroll to. I've used the id of the member as that should be unique.

File: src/main/java/org/snomed/snowstorm/rest/ControllerHelper.java [hunk 2]
```diff
@@ -133,6 +135,15 @@ public static PageRequest getPageRequest(int offset, int limit, Sort sort) {
...
 
+	public static PageRequest getPageRequest(int offset, int limit, Sort sort, String searchAfter) {
+		validatePageSize(offset, limit);
+		if (!Strings.isNullOrEmpty(searchAfter)) {
+			return SearchAfterPageRequest.of(SearchAfterHelper.fromSearchAfterToken(searchAfter), limit, sort);
+		} else {
+			return ControllerHelper.getPageRequest(offset, limit);
+		}
+	}
+
```
Comment (dmcgihtsdo) [inline src/main/java/org/snomed/snowstorm/rest/ControllerHelper.java:146]: searchAfter here is nullable, which allows for a minor change in the controller.

File: src/test/java/org/snomed/snowstorm/core/data/services/RefsetDescriptorUpdaterServiceTest.java [hunk 3]
```diff
@@ -89,7 +89,7 @@ void testUpdatingRefSetConcept() throws ServiceException {
 
 		// then
 		List<ReferenceSetMember> members = memberService.findMembers("MAIN", buildMemberSearchRequest(true, Concepts.REFSET_DESCRIPTOR_REFSET), PAGE_REQUEST).getContent();
-		ReferenceSetMember referenceSetMember = members.get(2); // newRefSet2 is 3rd document.
+		ReferenceSetMember referenceSetMember = getReferenceSetMember(members, newRefSet2.getConceptId());
```
Comment (dmcgihtsdo) [inline src/test/java/org/snomed/snowstorm/core/data/services/RefsetDescriptorUpdaterServiceTest.java:92]: This is not related to MAINT-1790, however, it is a temperamental test that occasionally failed. I've fixed it by explicitly asking for the member I want.

File: src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		// query second page (using searchAfter token from firstPage)
+		ItemsPagePojo<ReferenceSetMember> secondPageMembers = findRefsetMembers(firstPageMembers.getSearchAfter());
+		Thread.sleep(10000);
+		assertEquals(1, secondPageMembers.getItems().size());
+		ReferenceSetMember secondPageMember = getReferenceSetMember(secondPageMembers);
+		assertEquals(memberB.getMemberId(), secondPageMember.getMemberId());
+
+		// should be different, thus paged
+		assertNotEquals(firstPageMember, secondPageMember);
+	}
```
Comment (dmcgihtsdo) [inline src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java:65]: This essentially tests page 1 and page 2 of the endpoint.

File: src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+		// query first page (using offset & limit)
+		ItemsPagePojo<ReferenceSetMember> firstPageMembers = findRefsetMembers(0, 1);
+		assertEquals(1, firstPageMembers.getItems().size());
+		ReferenceSetMember firstPageMember = getReferenceSetMember(firstPageMembers);
+		assertEquals(memberA.getMemberId(), firstPageMember.getMemberId());
+
+		// query second page (using searchAfter token from firstPage)
+		ItemsPagePojo<ReferenceSetMember> secondPageMembers = findRefsetMembers(firstPageMembers.getSearchAfter());
+		Thread.sleep(10000);
```
Comment (dmcgihtsdo) [inline src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java:58]: The test would occasionally fail. I added in this sleep which resulted in the test consistently passing. I presume searchAfter is slower? I'm not sure. I've tried decreasing the sleep time but it didn't allow for the test to pass. Do you have any recommendations on a better approach? Sleeping for 10 seconds isn't ideal.

File: src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+		ReferenceSetMember secondPageMember = getReferenceSetMember(secondPageMembers);
+		assertEquals(memberB.getMemberId(), secondPageMember.getMemberId());
+
+		// should be different, thus paged
+		assertNotEquals(firstPageMember, secondPageMember);
+	}
+
+	private ReferenceSetMember getReferenceSetMember(ItemsPagePojo<ReferenceSetMember> firstPageMembers) {
+		return firstPageMembers.getItems().iterator().next();
+	}
```
Comment (dmcgihtsdo) [inline src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java:69]: Only 1 referenceSetMember is being returned at any given time (for my test) so it is safe to use iterator.

File: src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		referenceSetMemberService.createMember("MAIN", memberB);
+		assertEquals(2, findRefsetMembers().getItems().size()); // memberA & memberB
+
+		// query first page (using offset & limit)
+		ItemsPagePojo<ReferenceSetMember> firstPageMembers = findRefsetMembers(0, 1);
+		assertEquals(1, firstPageMembers.getItems().size());
+		ReferenceSetMember firstPageMember = getReferenceSetMember(firstPageMembers);
+		assertEquals(memberA.getMemberId(), firstPageMember.getMemberId());
+
+		// query second page (using searchAfter token from firstPage)
```
Comment (kaicode) [inline src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java:56]: Loading the second page using searchAfter will only work consistently if the sort order is also applied when not using searchAfter (when loading the first page).

**Review by kaicode — CHANGES_REQUESTED** [review]

File: src/main/java/org/snomed/snowstorm/rest/ControllerHelper.java [hunk 8]
```diff
@@ -133,6 +135,15 @@ public static PageRequest getPageRequest(int offset, int limit, Sort sort) {
 		return sort == null ? PageRequest.of(page, size) : PageRequest.of(page, size, sort);
 	}
 
+	public static PageRequest getPageRequest(int offset, int limit, Sort sort, String searchAfter) {
+		validatePageSize(offset, limit);
+		if (!Strings.isNullOrEmpty(searchAfter)) {
+			return SearchAfterPageRequest.of(SearchAfterHelper.fromSearchAfterToken(searchAfter), limit, sort);
+		} else {
+			return ControllerHelper.getPageRequest(offset, limit);
```
[the remark this answers is not in the document]
Comment (dmcgihtsdo) [reply src/main/java/org/snomed/snowstorm/rest/ControllerHelper.java]: Thanks, I have changed this.

File: src/main/java/org/snomed/snowstorm/rest/ReferenceSetMemberController.java [hunk 9]
```diff
@@ -41,6 +42,7 @@
 @Api(tags = "Refset Members", description = "-")
 @RequestMapping(produces = "application/json")
 public class ReferenceSetMemberController {
+	private static final Sort SORT_BY_MEMBER_ID_DESC = Sort.sort(ReferenceSetMember.class).by(ReferenceSetMember::getMemberId).descending();
```
[the remark this answers is not in the document]
Comment (dmcgihtsdo) [reply src/main/java/org/snomed/snowstorm/rest/ReferenceSetMemberController.java:46]: Thanks, I have changed this.

File: src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java (hunk 7 above)
Comment (dmcgihtsdo) [reply src/test/java/org/snomed/snowstorm/rest/ReferenceSetMemberControllerTest.java:60]: Good thinking. I have added a test for this.

**Review by kaicode — APPROVED** [review]
Looks good to me!
